backend/OKPluginVariableRendering/VariableRendering.cs

```diff
@@ -0,0 +1,358 @@
+                        {
+                            continue;
+                        }
+
+                        // check if attribute exists in base ci
+                        if (mainCI.MergedAttributes.ContainsKey(attribute.Name) && ciPrio < attribute.Priority)
+                        {
+                            // don't change this attribute
+                            continue;
+                        }
+
+                        fragments.Add(new BulkCIAttributeDataLayerScope.Fragment(GetTargetName(attribute.Name, mapping.Target), attribute.Value, mainCI.ID));
+
+                    }
+
+                }
+            }
+
+            await attributeModel.BulkReplaceAttributes(
+                new BulkCIAttributeDataLayerScope("", targetLayer.ID, fragments),
+                changesetProxy, 
+                new DataOriginV1(DataOriginType.ComputeLayer), 
+                trans,
+                MaskHandlingForRemovalApplyNoMask.Instance);
+
+            return true;
+        }
+
+
+        private static bool IsAttributeAllowed(string attribute, List<string> attributeWhitelist, List<string> attributeBlacklist)
+        {
+            var result = false;
+
+            if (attributeBlacklist.Count > 0 && attributeBlacklist[0] == "*")
+            {
+                // NOTE: in this case all attributes are on blacklist
+                return result;
+            }
+
+            if (attributeWhitelist.Count == 0 || attributeWhitelist[0] == "*")
+            {
+                // NOTE: in this case all attributes are allowed
+                result = true;
+            }
+
+            if (attributeWhitelist.Where(a => attribute.StartsWith(a[..1])).ToList().Count > 0)
+            {
+                result = true;
+            }
+
+            if (attributeBlacklist.Where(a => attribute.StartsWith(a[..1])).ToList().Count > 0)
+            {
+                result = false;
+            }
+
+            return result;
+        }
+
+        private static bool IsAttributeIncludedInSource(string attribute, string source)
+        {
+            var result = false;
+
+            if (attribute == "__name")
+            {
+                //result = true;
+                return result;
+            }
+
+            if (source == "*")
+            {
+                result = true;
+            }
+            else if (attribute.StartsWith(source[..^1]))
+            {
+                result = true;
+            }
+
+            return result;
+        }
+
+        private static string GetTargetName(string source, string target)
+        {
+            if (source == "__name")
+            {
+                return source;
+            }
+
+            return target.Replace("{SOURCE}", source);
+        }
+
+        internal class GatheredAttribute
+        {
+            public Guid SourceCIID { get; set; }
+            public string RequiredTrait { get; set; }
```

Comment:
I don't think that's needed, is it?

Reply:
This is helping me with the gathering of attributes since I also need to take into consideration priorities of each attribute.